p2p/protocols/protocol.go

```diff
@@ -322,23 +339,40 @@ func (p *Peer) handleIncoming(handle func(ctx context.Context, msg interface{})
 		return errorf(ErrDecode, "<= %v: %v", msg, err)
 	}
 
-	// if the accounting hook is set, call it
+	// if the accounting hook is set, do accounting logic
 	if p.spec.Hook != nil {
-		err := p.spec.Hook.Receive(p, uint32(len(msgBytes)), val)
+
+		p.lock.Lock()
+		defer p.lock.Unlock()
+
+		size := uint32(len(msgBytes))
+
+		// validate that the accounting call would succeed...
+		costToLocalNode, err := p.spec.Hook.Validate(p, size, val, Receiver)
 		if err != nil {
+			// ...because if it would fail, we return and don't handle the message
 			return err
 		}
-	}
 
-	// call the registered handler callbacks
-	// a registered callback take the decoded message as argument as an interface
-	// which the handler is supposed to cast to the appropriate type
-	// it is entirely safe not to check the cast in the handler since the handler is
-	// chosen based on the proper type in the first place
-	if err := handle(ctx, val); err != nil {
-		return errorf(ErrHandler, "(msg code %v): %v", msg.Code, err)
+		// seems like accounting would be fine, so handle the message
+		if err := handle(ctx, val); err != nil {
```

Comment:
If the message handler is holding `p.lock` and sending also tries to acquire `p.lock` doesn't this mean that if a protocol uses `Peer.send` in one of its message handlers this could lead to a deadlock?

Reply:
Nevermind, this is probably not an issue if the handler routines are go routines which will be enforced soon anyways.

Reply:
Indeed. Also, the deadlock occurs if one routine acquires the lock and doesn't release it for some reason. Or for example, if something is sent to a channel and no routine reads from that channel (or a routine is waiting on a channel and nobody ever writes to that channel).

In this case above though, in that situation what should happen is that one "thread" (let's assume sending) first gets the lock, which means that if a concurrent receive happens, then the peer will be blocked until the sending "thread" did not finish. After it finishes, it should release the lock, at which point the receiving "thread" will obtain the lock and continues operation.

We have this situation right now anyway, just at the `Swap` instance, in that `Add` is both called by sending and receiving, thus one has to wait until the other finishes in case of concurrency.